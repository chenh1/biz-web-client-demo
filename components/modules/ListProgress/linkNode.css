@keyframes glow {
  0% {
    color: #F3F4F6;
  }
  100% {
    color: #AB46FF;
  }
}

@keyframes sway {
  0% {
    transform: translateX(10px);
  }
  100% {
    transform: translateX(0px);
  }

}

.link-node-text {
  animation: glow 1s linear infinite alternate;
}

.attention-chevron {
  animation: sway 0.5s linear infinite alternate;
}