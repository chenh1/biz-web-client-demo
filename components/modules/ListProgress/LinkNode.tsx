'use client'

import type { BgColor } from "@/components/containers/FixedDrawer/FixedDrawer";
import React, { useState } from "react";
import { Box } from "../../containers/Box/Box";
import { useRouter } from "next/navigation";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import './linkNode.css'
import '../../core/Button/button.css'

interface LinkNodeProps{
  title: string
  isCompleted: boolean;
  module: `${number}.${number}`;
  borderColor: `border-${BgColor}`;
  bgColor: `bg-${BgColor}`;
  guideToFirstNode: boolean;
}

export const LinkNode: React.FC<LinkNodeProps> = ({ title, isCompleted, module, borderColor, bgColor, guideToFirstNode }: LinkNodeProps) => {
  const router = useRouter();
  const [ isHover, setIsHover ] = useState<boolean>(false);
  
  return (
    <Box onClick={() => router.push(`/module/${module.replace('.', '-')}`)} onMouseEnter={() => setIsHover(true)} onMouseLeave={() => setIsHover(false)} padding="0" className="z-10 relative cursor-pointer grid-cols-[16px,_auto] items-center">
      {guideToFirstNode &&
        <Box padding="0" className="attention-chevron absolute">
          <div className="triangle static left">
            <div></div>
          </div>
        </Box>
      }
      <div className={`relative w-4 h-4 border-2 ${borderColor} rounded-full ${isHover || isCompleted || guideToFirstNode ? bgColor : 'bg-night-black'}`}>
        {isHover || isCompleted || guideToFirstNode && <div className={`origin-center absolute -top-0.5 -left-0.5 w-4 h-4 border-2 rounded-full ${borderColor} ${bgColor} animate-pulseEffect`}></div>}
      </div>
      {title && <Paragraph className={`${guideToFirstNode ? 'font-bold link-node-text' : ''}`} darkTextColor={isHover ? "apex-blue" : undefined} textColor={isHover ? "apex-blue" : undefined}>{`${guideToFirstNode ? 'START HERE: ' : ''}${title}`}</Paragraph>}
      {/* <Heading fontSizeClass="text-lg">{title}</Heading> */}
    </Box>
  )
}