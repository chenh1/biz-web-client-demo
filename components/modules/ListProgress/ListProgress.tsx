'use client'

import type { VideoType } from "@/graphql/queries/videos";
import type { ProgressData } from "@/graphql/queries/myData";
import type { RawTree } from "@/graphql/queries/learningPaths";
import type { BgColor } from "@/components/containers/FixedDrawer/FixedDrawer";
import React from "react";
import { Box } from "../../containers/Box/Box";
import { useParams } from "next/navigation";
import { buildLinearTree, formatTree } from "../../utils/treeDataUtils";
import { Heading } from "../../core/Heading/Heading";
import { LinkNode } from "./LinkNode";
import { FadeScroller } from "../../containers/FadeScroller/FadeScroller";


interface Video {
  attributes: VideoType;
}

interface ListProgressProps {
  isVertical?: boolean;
  highlightValue?: number;
  background?: string;
  showAsList?: boolean;
  nodeWidth?: number;
  nodeHeight?: number;
  rawTree: RawTree;
  faderColor?: string;
  twFaderColor?: string;
  allLearningPaths: {
    title: string;
    level: string;
    description: string;
    rawTree: RawTree;
  }
  className?: string;
  nameMap?: Record<string, string>;
  nameMapCallback?: (nameMap: Record<string, string> | undefined, name: string) => string;
  progressData: ProgressData;
  availableModules: Video[];
  dataCy: string;
  globalOnMouseEnter?: (e: any) => void;
}

// recursively dig into tree's nested nodes to create a flattened array of nodes in order
const mapTree = (tree: RawTree, userData: ProgressData): { [key: string]: { name: `${number}.${number}`; isCompleted: boolean; }[] } => {
  const completedModules = userData?.completedModules || [];
  const nodes = {};
  const traverse = (node) => {
    if (node.children) {
      node.children.forEach(child => {
        traverse(child);
      });
    }

    if (node.name) {
      node.isCompleted = completedModules.filter(({ lessonModule: lm }) => lm == node.name)?.length > 0
    }

    const mappedNode = {...node}
    mappedNode.children = null;
    const nodeCategoryKey = "s" + (parseInt(mappedNode?.name) - 1)

    if (isNaN(parseInt(mappedNode?.name))) {

    } else if (nodes[nodeCategoryKey]) {
      nodes[nodeCategoryKey].push(mappedNode)
    } else {
      nodes[nodeCategoryKey] = [mappedNode]
    }
  }

  traverse(tree);
  return nodes;
}

const borderKeys: `border-${BgColor}`[] = [
  'border-hyper-red',
  'border-force-yellow',
  'border-ultra-purple',
  'border-cyber-teal',
  'border-apex-blue'
]

const bgKeys: `bg-${BgColor}`[] = [
  'bg-hyper-red',
  'bg-force-yellow',
  'bg-ultra-purple',
  'bg-cyber-teal',
  'bg-apex-blue'
]

export const ListProgress: React.FC<ListProgressProps> = (props: ListProgressProps) => {
  const { rawTree: learningPathTree, twFaderColor, faderColor, allLearningPaths = {}, nameMap, nameMapCallback, progressData, availableModules } = props || {}
  
  const rawTree = learningPathTree ? formatTree(learningPathTree, progressData) : buildLinearTree(availableModules, progressData);
  const mappedTree = mapTree(JSON.parse(JSON.stringify(rawTree)), progressData)
  const mapKeys = Object.keys(mappedTree)?.reverse()
  const guideToFirstNode = progressData?.completedModules?.length === 0 || !progressData?.completedModules
  
  return (
    <FadeScroller deps={[ progressData, learningPathTree, allLearningPaths ]} orientation="vertical" className="max-h-[70vh] sm:max-h-[80vh]" twFaderColor={twFaderColor} faderColor={faderColor}>
      <Box padding="0" className="py-12">
        {mapKeys?.map((key, i) => {
          const nodes = mappedTree[key]?.reverse()
          const index = parseInt(key.replace('s', ''));
          const { attributes: { title } = {} } = allLearningPaths[index] || {}

          return (
            <Box key={i}>
              <Heading type="h4" fontSizeClass="text-lg">{title}</Heading>
              <Box padding="0" gap="sm" className="">
                {nodes.map((node, j) => {
                  const title = nameMapCallback ? nameMapCallback(nameMap, node.name) : ""
                  return (
                    <div key={j} className="relative">
                      {j !== 0 && <div className={`absolute border ${borderKeys[index]} h-[48px] w-px left-2 -translate-y-[32px] -translate-x-1/2`}></div>}
                      <LinkNode guideToFirstNode={guideToFirstNode && i === 0 && j === 0} bgColor={bgKeys[index]} borderColor={borderKeys[index]} isCompleted={node.isCompleted} title={title} module={node.name}/>
                    </div>
                  )
                })}
              </Box>
            </Box>
          )
        })}
      </Box>
    </FadeScroller>
  )
}