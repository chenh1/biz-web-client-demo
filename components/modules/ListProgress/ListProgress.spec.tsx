import { ListProgress } from "./ListProgress";
import { render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders ListProgress module', async () => {
  const { getByTestId, asFragment } = render(
    <ListProgress
      rawTree={{
        name: "level 1", isCompleted: false, children: [] 
      }}
      allLearningPaths={{
        title: "Intro to cooking", level: "", description: "", rawTree: { name: "level 1", isCompleted: false, children: [] } 
      }}
      progressData={{
        currentModule: {
          lessonModule: "1.01"
        },
        completedModules: [
          { lessonModule: "1.01", examScore: 100 },
          { lessonModule: "1.02", examScore: 100 }
        ]
      }}
      availableModules={[
        { attributes: { title: "Intro to cooking", url: "https://www.test.com", description: "test", transcript: "test", lessonModule: 1.01, tags: { data: [{ attributes: { displayName: "Tag 1", color: "blue", value: "tag1" }}] } }},
        { attributes: { title: "Intermediate cooking", url: "https://www.test.com", description: "test", transcript: "test", lessonModule: 1.02, tags: { data: [{ attributes: { displayName: "Tag 1", color: "blue", value: "tag1" }}] } }},
        { attributes: { title: "Master chef", url: "https://www.test.com", description: "test", transcript: "test", lessonModule: 1.03, tags: { data: [{ attributes: { displayName: "Tag 1", color: "blue", value: "tag1" }}] } }}
      ]}
      dataCy="test"
    />
  )

  expect(asFragment()).toMatchSnapshot();
});
