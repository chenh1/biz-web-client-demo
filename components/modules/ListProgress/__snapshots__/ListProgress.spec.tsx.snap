// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ListProgress module 1`] = `
<DocumentFragment>
  <div
    class=" grid  gap-4  p-2 relative max-h-[70vh] sm:max-h-[80vh]"
  >
    <div
      class=" grid  gap-4  p-2 fade-scroller max-w-full overflow-y-auto overflow-x-hidden"
    >
      <div
        class=" grid  gap-4  p-2 py-12"
      />
    </div>
  </div>
</DocumentFragment>
`;
